### PR TITLE
Fix for prior fix.

### PR DESCRIPTION
**Pull request by chfritz**
Sorry, there was an issue with the previous attempt of fixing this.
This one seems to work.

### CODE REVIEW

**Comment by chris-smith** [comment]
Did the previous commit just have an issue with comments? ROS seems to use the entire line after the equals sign (up to a comment if it's not a string) to calculate the md5 text - so (`float32 FLOAT_CONST=1.0` gives a different md5sum than `float32 FLOAT_CONST=1.00`)

I had to do something similar in the [indigo branch](https://github.com/chris-smith/rosjs/blob/indigo-devel/utils/messageGeneration/MessageSpec.js#L247) actually - ended up consulting `genmsg` source code a lot. Mainly [gentools](https://github.com/ros/genmsg/blob/490b3753213ed3e2251acd218c62224b2787ee0e/src/genmsg/gentools.py#L59), [msgs](https://github.com/ros/genmsg/blob/490b3753213ed3e2251acd218c62224b2787ee0e/src/genmsg/msgs.py#L158), and [msg_loader](https://github.com/ros/genmsg/blob/490b3753213ed3e2251acd218c62224b2787ee0e/src/genmsg/msg_loader.py#L183)

**Comment by chfritz** [comment]
Thanks for the pointer. That is indeed a better way of dealing with the issue and I've adopted that now, but I still needed to store the constant value as a string, rather than a parsed value in order to stop javascript from doing things like turning 1.0 into 1.

**Comment by jeffadamsc** [comment]
Do you still have a problem with  "1.0" vs "1.00" in this change? I see you're looking for whitespace, but not trailing zeros.

File: utils/messages.js [hunk 1]
```diff
@@ -341,14 +341,31 @@ function extractFields(content, details, callback) {
...
+              constantEnd = constantComment;
+            }
+          }
           var parsedConstant = fieldsUtil.parsePrimitive(fieldType, constant);
 
           constants.push({
             name        : fieldName
             , type        : fieldType
             , value       : parsedConstant
+<<<<<<< HEAD
```
Comment (chris-smith) [inline utils/messages.js]: merge issue

File: utils/messages.js [hunk 2]
```diff
@@ -504,8 +521,7 @@ function buildMessageClass(details) {
             new (field.messageType)(values ? values[field.name] : undefined);
         } else {
           // simple value
-          that[field.name] = values ? values[field.name] :
-            (field.value || fieldsUtil.getDefaultValue(field.type));
+          that[field.name] = values ? values[field.name] : null;
```
Comment (chris-smith) [inline utils/messages.js]: also a merge issue?

**Comment by chris-smith** [comment]
> Do you still have a problem with "1.0" vs "1.00" in this change? I see you're looking for whitespace, but not trailing zeros.

1.0 is different from 1.00 in ros' message md5 text so trailing zeros are correct.

File: utils/messages.js (hunk 1 above)
Comment (chfritz) [reply utils/messages.js]: oops, sorry. fixed now.

File: utils/messages.js (hunk 2 above)
Comment (chfritz) [reply utils/messages.js]: no this one was intended.

Comment (chris-smith) [reply utils/messages.js]: Won't this force users to specify values for all the fields in their messages?

```
const header = new std_msgs.msg.Header()
pub.publish(header); // invalid
```

whereas before that would have worked

**Comment by chfritz** [comment]
I'm actually not sure what the right behavior is. We also want to allow users to leave (optional) values unspecified (null). Maybe we just need to set them explicitly to null then already in the JS object.

**Comment by chris-smith** [comment]
I don't think ROS 1.0's message serialization scheme allows for optional values.

If we want optional arguments when sending messages we typically will use constants or something and check against those on the other end.

```
if (msg.optionalField !== MyMessage.OPTIONAL_DEFAULT) { ... }
// or
if (msg.optionalList.length > 0) { ... }
```

It does start to get cumbersome when you want optional complex fields.

Automatically providing default values for fields allows devs sending a message to ignore certain fields if they don't care about them though - provided your subscriber interprets the default values correctly...

**Comment by chfritz** [comment]
I guess you are right. I was thrown off by some null values I saw in our messages, but I guess those are just used as constants the same way you guys are using them.

Fixed that now, allowing for null values but providing defaults for fields that are actually `undefined`, i.e., missing in the goal object.

**Review by chris-smith — APPROVED** [review]

**Comment by chris-smith** [comment]
Awesome - I'll release a new version with this and then pull the indigo merge changes in.

**Pull request merged**